File: Coding_1215 1D_DP/최원준/이강훈/3번.java
/*
https://leetcode.com/problems/house-robber-ii/description/

아이디어 ( 솔루션 보고 품 )
2차원 DP
 -> [start][end] 팰린드롬 가능 여부 저장
  -> 앞 자리부터 2중 for 문으로 팰린드롬 체크
   -> [start][end] 팰린드롬 조건은 charAt[start]==charAt[end] && ( end-start < 2 || [start+1][end-1]==true )

??
https://leetcode.com/problems/longest-palindromic-substring/solutions/4212564/beats-96-49-5-different-approaches-brute-force-eac-dp-ma-recursion/
Approach 4: Manacher's Algorithm
이건 O(N) 인데 무슨 방법인지..

자료구조
2차원 배열

시간복잡도
O (N^2)
*/

public class Solution {
    public String longestPalindrome(String s) {
        boolean[][] dy = new boolean[s.length()][s.length()];

        int maxLength = Integer.MIN_VALUE;
        int start = 0;
        int end = 0;

        for (int i = 0; i < s.length(); i++) {
            dy[i][i] = true;

            for (int j = 0; j < i; j++) {
                if (s.charAt(i) == s.charAt(j) && (i - j < 2 || dy[j+1][i-1])) {
                    dy[j][i] = true;

                    if ((i - j) + 1 > maxLength) {
                        maxLength = (i - j) + 1;
                        start = j;
                        end = i;
                    }
                }
            }
        }

        return s.substring(start, end + 1);
    }
}

File: Coding_231215 1D_DP/최원준/최원준/3번.py
# https://leetcode.com/problems/longest-palindromic-substring/description/


class Solution:
    def longestPalindrome(self, s: str) -> str:
        if s == s[::-1]:
            return s
        size = 1
        start = 0
        for idx in range(1, len(s)):
            left = idx - size
            right = idx + 1
            big = s[left - 1:right]
            lil = s[left:right]
            print(big,"/", lil)
            if big == big[::-1] and len(big) > size:
                size += 2
                start = left - 1
            elif lil == lil[::-1] and len(lil) > size:
                size += 1
                start = left

        return s[start:start+size]
